src/goto-instrument/contracts/dynamic-frames/dfcc_dsl_wrapper_program.cpp

```diff
@@ -0,0 +1,1141 @@
+  is_fresh.rewrite_calls(requires_program, address_of_requires_write_set);
+
+  // rewrite is_freeable predicates
+  dfcc_is_freeablet is_freeable(library, log);
+  is_freeable.rewrite_calls(requires_program, address_of_requires_write_set);
+
+  // instrument for side effects
+  instrument.instrument_goto_program(
+    wrapper_id, requires_program, address_of_requires_write_set);
+
+  // append resulting program to preconditions section
+  preconditions.destructive_append(requires_program);
+  log.debug() << "<-dfcc_dsl_wrapper_programt::encode_requires_clauses()"
+              << messaget::eom;
+}
+
+void dfcc_dsl_wrapper_programt::encode_ensures_clauses()
+{
+  log.debug() << "->dfcc_dsl_wrapper_programt::encode_ensures_clauses()"
+              << messaget::eom;
+  // we need the contract write set for the freed predicates
+  PRECONDITION(addr_of_contract_write_set_symbol_opt);
+  // we need the ensures write set to check for side effects
+  PRECONDITION(addr_of_ensures_write_set_symbol_opt);
+  const auto &language_mode = wrapper_symbol.mode;
+  const auto &wrapper_id = wrapper_symbol.name;
+
+  code_contractst code_contracts(goto_model, log);
+
+  // Build the ensures clause program and snapshot program
+  exprt::operandst ensures_conjuncts;
+  for(const auto &r : contract_code_type.ensures())
+  {
+    ensures_conjuncts.push_back(
+      to_lambda_expr(r).application(contract_lambda_parameters));
+  }
+
+  source_locationt sl =
+    ensures_conjuncts.empty()
+      ? contract_code_type.source_location()
+      : contract_code_type.ensures().front().source_location();
+
+  sl.set_function(wrapper_symbol.name);
+
+  if(contract_mode == dfcc_contract_modet::CHECK)
+  {
+    sl.set_property_class(ID_postcondition);
+    sl.set_comment(
+      "Check ensures clause of contract " + id2string(contract_symbol.name) +
+      " for function " + id2string(wrapper_id));
+  }
+
+  auto ensures = conjunction(ensures_conjuncts);
+
+  if(has_subexpr(ensures, ID_exists) || has_subexpr(ensures, ID_forall))
+    code_contracts.add_quantified_variable(ensures, language_mode);
+
+  // Generate pair of programs from the ensures statement where:
+  // - the first program takes snapshots of the history variables
+  goto_programt history_snapshot_program;
+
+  // this also rewrites ID_old expressions to fresh variables
+  code_contracts.generate_history_variables_initialization(
+    ensures, language_mode, history_snapshot_program);
+
+  // - the second program is the ensures clause converted to GOTO with history
+  // variables replaced by snapshot variables
+  const auto &statement_type =
+    (contract_mode == dfcc_contract_modet::CHECK) ? ID_assert : ID_assume;
+
+  codet ensures_statement(statement_type, {std::move(ensures)});
+  goto_programt ensures_program;
+  converter.goto_convert(ensures_statement, ensures_program, language_mode);
+
+  // inline all calls in the preconditions
+
+  ensures_program.instructions.back().source_location_nonconst() = sl;
```

Comment:
As above: do this before calling `goto_convert`.

Reply:
fixed